msal-go/error-design.md

```diff
@@ -0,0 +1,108 @@
+---
+title: Microsoft Authentication Library for Go Error Design
+description: "Error design guidelines for the Microsoft Authentication Library for Go."
+---
+
+# Microsoft Authentication Library for Go Error Design
+
+## Background
+
+Errors in MSAL are intended for app developers to troubleshoot and not for displaying to end-users.
+
+### Go error handling vs. other languages
+
+Most modern languages use exception-based errors. Simply put, you "throw" an exception and it must be caught at some routine in the upper stack or it will eventually crash the program.
+
+Go doesn't use exceptions, instead it relies on multiple return values, one of which can be the builtin error interface type. It is up to the user to decide what to do.
+
+### Go custom error types
+
+Errors can be created in Go by using `errors.New()` or `fmt.Errorf()` to create an "error".
+
+Custom errors can be created in multiple ways. One of the more robust ways is simply to satisfy the error interface:
+
+```go
+type MyCustomErr struct {
+  Msg string
+}
+func (m MyCustomErr) Error() string { // This implements "error"
+  return m.Msg
+}
+```
+
+### MSAL error goals
```

Comment:
Thoughts on placing these points at the top of the article, - just below the sentence in ##background?